test/jdk/java/nio/channels/DatagramChannel/Connect.java

```diff
@@ -43,89 +47,90 @@ public static void main(String[] args) throws Exception {
     }
 
     static void test() throws Exception {
-        Reactor r = new Reactor();
-        Actor a = new Actor(r.port());
-        invoke(a, r);
+        ExecutorService threadPool = Executors.newCachedThreadPool();
+        try (Reactor r = new Reactor();
+             Actor a = new Actor(r.port())) {
+            invoke(threadPool, a, r);
+        } finally {
+            threadPool.shutdown();
+        }
     }
 
-    static void invoke(Sprintable reader, Sprintable writer) throws Exception {
-
-        Thread writerThread = new Thread(writer);
-        writerThread.start();
-
-        Thread readerThread = new Thread(reader);
-        readerThread.start();
-
-        writerThread.join();
-        readerThread.join();
-
-        reader.throwException();
-        writer.throwException();
+    static void invoke(ExecutorService e, Runnable reader, Runnable writer) throws CompletionException {
+        CompletableFuture<Void> f1 = CompletableFuture.runAsync(writer, e);
+        CompletableFuture<Void> f2 = CompletableFuture.runAsync(reader, e);
+        wait(f1, f2);
     }
 
-    public interface Sprintable extends Runnable {
-        public void throwException() throws Exception;
+
+    // This method waits for either one of the given futures to complete exceptionally
+    // or for all of the given futures to complete successfully.
+    private static void wait(CompletableFuture<?>... futures) throws CompletionException {
+        CompletableFuture<?> future = CompletableFuture.allOf(futures);
+        Stream.of(futures)
+                .forEach(f -> f.exceptionally(ex -> {
+                    future.completeExceptionally(ex);
+                    return null;
+                }));
+        future.join();
     }
 
-    public static class Actor implements Sprintable {
+    public static class Actor implements AutoCloseable, Runnable {
         final int port;
-        Exception e = null;
+        final DatagramChannel dc;
 
-        Actor(int port) {
+        Actor(int port) throws IOException {
```

Comment:
I think it would be clear to create Actor with a SocketAddress rather than a port. It didn't matter in the old test because the Reactor was bound to the wildcard address but here there is nothing to tell the Actor which address to connect it.

Reply:
Thanks for the feedback Alan! Will look into these comments presently